backend/optimize.ml

```diff
@@ -0,0 +1,250 @@
+    let fst_val = DLL.value fst in
+    let snd_val = DLL.value snd in
+    match fst_val.desc with
+    | Op (Move | Spill | Reload) -> (
+      let fst_src, fst_dst = fst_val.arg.(0), fst_val.res.(0) in
+      match snd_val.desc with
+      | Op (Move | Spill | Reload) ->
+        let snd_src, snd_dst = snd_val.arg.(0), snd_val.res.(0) in
+        if are_equal_regs fst_src snd_dst && are_equal_regs fst_dst snd_src
+        then (
+          if !Flambda_backend_flags.cfg_peephole_optimize_track
+          then update_csv "remove_useless_mov";
+          DLL.delete_curr snd;
+          Some (prev_at_most go_back_const fst))
+        else None
+      | _ -> None)
+    | _ -> None)
+  | _ -> None
+
+(** Logical condition for simplifying the following case: 
+  {| 
+    <op 1> const1, r 
+    <op 2> const2, r 
+  |}
+
+  to: 
+  {|
+    <op 1> (const1 <op 2> const2), r 
+  |}
+
+   Where <op 1> and <op 2> can be any two binary operators that are associative
+   and const1 and const2 are immediate values. *)
+
+(* CR-soon gtulba-lecu for gtulba-lecu: implement the rest of Intop_imm cases
+   (only Imulh left to optimize if possible)
+
+   it may be the case that after Iadd, Isub folding the result is 0, treat this
+   case. *)
+let are_compatible op1 op2 imm1 imm2 =
+  match (op1 : Mach.integer_operation), (op2 : Mach.integer_operation) with
+  | Mach.Iand, Mach.Iand -> Some (Mach.Iand, imm1 land imm2)
+  | Ior, Ior -> Some (Mach.Ior, imm1 lor imm2)
+  | Ixor, Ixor -> Some (Mach.Ixor, imm1 lxor imm2)
+  | Ilsl, Ilsl ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Ilsl, imm1 + imm2)
+    else None
+  | Ilsr, Ilsr ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Ilsr, imm1 + imm2)
+    else None
+  | Iasr, Iasr ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Iasr, imm1 + imm2)
+    else None
+  | Iadd, Iadd ->
+    if Misc.no_overflow_add imm1 imm2
+    then Some (Mach.Iadd, imm1 + imm2)
+    else None
+  | Iadd, Isub ->
+    if imm1 >= imm2
+    then Some (Mach.Iadd, imm1 - imm2)
+    else Some (Mach.Isub, imm2 - imm1)
+  | Isub, Isub ->
+    if Misc.no_overflow_add imm1 imm2
+    then Some (Mach.Isub, imm1 + imm2)
+    else None
+  | Isub, Iadd ->
+    if imm1 >= imm2
+    then Some (Mach.Isub, imm1 - imm2)
+    else Some (Mach.Iadd, imm2 - imm1)
+  | Ilsl, Imul ->
+    if Misc.no_overflow_mul (1 lsl imm1) imm2
+    then Some (Mach.Imul, (1 lsl imm1) * imm2)
+    else None
+  | Imul, Ilsl ->
+    if Misc.no_overflow_mul imm1 (1 lsl imm2)
+    then Some (Mach.Imul, imm1 * (1 lsl imm2))
+    else None
+  | Imul, Imul ->
+    if Misc.no_overflow_mul imm1 imm2
+    then Some (Mach.Imul, imm1 * imm2)
+    else None
+  | Idiv, Idiv ->
+    if Misc.no_overflow_mul imm1 imm2
+    then Some (Mach.Idiv, imm1 * imm2)
+    else None
+  | Imod, Imod -> if imm1 mod imm2 = 0 then Some (Mach.Imod, imm2) else None
```

Comment:
Similarly, the following function should
return the results before/after optmization:
```
let imod (x : int64) (imm0 : int) (imm1 : int) : int64 * int64 =
  let open Int64 in
  rem (rem x (of_int imm0)) (of_int imm1),
  rem x (of_int imm1);;
```
and `imod 259L 3 5` returns:
```
(1L, 4L)
```

Reply:
Wait, my example does not make
any sense: `3 mod 5` is not zero.